### PR TITLE
Merge allOf

### PR DESCRIPTION
**Pull request by reuvenharrison**
## Objective
This PR introduces a function that simplifies OpenAPI specifications by removing subschemas under `allOf` and replacing them with a single merged (or flattened) top-level schema.
The resulting schema remains equivalent to the original one in terms of validation.
The use case we had in mind for this PR is for detecting breaking-changes which is implemented in https://github.com/Tufin/oasdiff using kin-openapi3. There may be other use cases which we did not think about.

## The `Merge` function
* The `Merge` function accepts a schema as input and returns a new equivalent schema without `allOf` 
* If the specification can't be merged, `Merge` returns an error
* Source: openapi3/schema_merge.go

## Unit testing
* Test how `Merge` works for different kinds of subschemas under `allOf`
* Source: openapi3/schema_merge_test.go

## Validation testing
* In order to verify equivalence of the original schema and the merged schema, we apply `ValidateRequest` to various inputs and verify that either the input matches both schemas or the input doesn't match both schemas, if one matches and the other doesn't, then the schemas are not equivalent.
* Source: openapi3filter/validation_merge_allof_test.go

## Development Status
* This PR is still under active development, and not all schema fields are currently supported.
* At present, we support the following schema fields:
  * Properties
  * AdditionalProperties
  * MinProps
  * MaxProps
  * Type
  * Format
  * Required
  * Min
  * Max
  * ExclusiveMin
  * ExclusiveMax
  * Items
  * MinItems
  * MaxItems
  * UniqueItems
  * Enum
  * MinLength
  * MaxLength
  * MultipleOf
  * Pattern: not supported because combining regex expressions requires positive lookahead which isn't supported by kin-openapi's regex engine
  * Title
  * Description
  * Not
  * OneOf and AnyOf: supported, but there might be some tricky situations where the result doesn't quite match the original specification, we are working to resolve these cases.

### CODE REVIEW

File: openapi3/schema_merge.go [hunk 1]
```diff
@@ -0,0 +1,789 @@
...
+	}
+	return findIntersectionOfArrays(nonEmptyEnum)
+}
+
+func resolvePattern(values []string) string {
+	var pattern strings.Builder
+	for _, p := range values {
+		if len(p) > 0 {
+			if !isPatternResolved(p) {
+				pattern.WriteString(fmt.Sprintf("(?=%s)", p))
```
Comment (reuvenharrison) [inline openapi3/schema_merge.go:439]: @tcdsv can you explain how this combines the patterns?

Comment (tcdsv) [reply openapi3/schema_merge.go:439]: we use positive lookahead (?=<pattern>) to specify that regex patterns must follow eachother, this enabled to combine multiple patterns.

The part you're talking about helps us avoid double-wrapping a pattern with "(?=..)"
This suggests a refactor for making that intent clearer.
Also, I've noticed that we don't have any validation tests for the pattern, I will add it.

File: openapi3/schema_merge_test.go [hunk 2]
```diff
@@ -0,0 +1,1033 @@
...
+				Value: &Schema{
+					Type:         "object",
+					ExclusiveMax: true,
+					Max:          Float64Ptr(1),
+				},
+			},
+			&SchemaRef{
+				Value: &Schema{
+					Type:         "object",
+					ExclusiveMin: false,
```
Comment (reuvenharrison) [inline openapi3/schema_merge_test.go:24]: Note this is ExclusiveMin.
Perhaps you meant ExclusiveMax?

File: openapi3/schema_merge_test.go [hunk 3]
```diff
@@ -0,0 +1,1033 @@
...
+				Value: &Schema{
+					Type:         "object",
+					ExclusiveMax: false,
+					Max:          Float64Ptr(1),
+				},
+			},
+			&SchemaRef{
+				Value: &Schema{
+					Type:         "object",
+					ExclusiveMin: true,
```
Comment (reuvenharrison) [inline openapi3/schema_merge_test.go:48]: Note this is ExclusiveMin.
Perhaps you meant ExclusiveMax?

File: openapi3/schema_merge_test.go (hunk 2 above)
Comment (tcdsv) [reply openapi3/schema_merge_test.go:24]: fixed, thanks.

File: openapi3/schema_merge_test.go [hunk 4]
```diff
@@ -0,0 +1,1273 @@
...
+								Required: []string{"boolean"},
+							},
+						},
+					},
+				},
+			},
+		},
+	})
+	require.NoError(t, err)
+	require.Equal(t, []string{"string", "boolean"}, merged.OneOf[0].Value.Required)
```
Comment (reuvenharrison) [inline openapi3/schema_merge_test.go:310]: I think it should be `boolean` only.

File: openapi3/schema_merge_test.go [hunk 5]
```diff
@@ -0,0 +1,1273 @@
...
+							},
+						},
+					},
+					Type: "object",
+				},
+			},
+		},
+	})
+
+	require.EqualError(t, err, TypeErrorMessage)
```
Comment (reuvenharrison) [inline openapi3/schema_merge_test.go:140]: This error is used for two different scenarios.
I recommend to use a dedicated error here.

File: openapi3/schema_merge.go [hunk 6]
```diff
@@ -0,0 +1,887 @@
...
+		return schema, nil
+	}
+
+	if areFormatsNumeric(formats) {
+		orderMap := make(map[string]int)
+		orderMap[formatInt32] = 1
+		orderMap[formatInt64] = 2
+		orderMap[formatDouble] = 3
+		orderMap[formatFloat] = 4
+		result := formatFloat
```
Comment (reuvenharrison) [inline openapi3/schema_merge.go:543]: Shouldn't the order be reversed?
According to the OpenAPI Spec, Double is wider than Float:
| Type | Format | Description|
|-- | -- | -- |
|number | float | Floating-point numbers.|
|number | double | Floating-point numbers with double precision.|

File: openapi3/schema_merge.go [hunk 7]
```diff
@@ -0,0 +1,887 @@
...
+	if len(formats) == 0 {
+		schema.Format = ""
+		return schema, nil
+	}
+
+	if areFormatsNumeric(formats) {
+		orderMap := make(map[string]int)
+		orderMap[formatInt32] = 1
+		orderMap[formatInt64] = 2
+		orderMap[formatDouble] = 3
```
Comment (reuvenharrison) [inline openapi3/schema_merge.go:541]: Shouldn't the order be reversed?
According to the OpenAPI Spec, Double is wider than Float:
| Type | Format | Description|
|-- | -- | -- |
|number | float | Floating-point numbers.|
|number | double | Floating-point numbers with double precision.|

File: openapi3/schema_merge_test.go [hunk 8]
```diff
@@ -0,0 +1,1273 @@
...
+					Title: "second",
+				},
+			},
+		},
+	})
+	require.NoError(t, err)
+	require.Equal(t, "base schema", merged.Title)
+}
+
+func TestMerge_NumericFormat(t *testing.T) {
```
Comment (reuvenharrison) [inline openapi3/schema_merge_test.go:788]: I recommend to split this into 3 tests

File: openapi3/schema_merge_test.go [hunk 9]
```diff
@@ -0,0 +1,1273 @@
...
+		},
+	})
+	require.EqualError(t, err, FormatErrorMessage)
+}
+
+func TestMerge_EmptySchema(t *testing.T) {
+	schema := Schema{}
+	merged, err := Merge(schema)
+	require.NoError(t, err)
+	require.Equal(t, &schema, merged) //todo &schema
```
Comment (reuvenharrison) [inline openapi3/schema_merge_test.go:940]: I suggest to clarify the TODO message or remove it

File: openapi3/schema_merge_test.go [hunk 10]
```diff
@@ -0,0 +1,1273 @@
...
+}
+
+func TestMerge_NoAllOf(t *testing.T) {
+	schema := Schema{
+		Title: "test",
+	}
+	merged, err := Merge(schema)
+	require.NoError(t, err)
+	require.Equal(t, &schema, merged) //todo &schema
```
Comment (reuvenharrison) [inline openapi3/schema_merge_test.go:949]: I suggest to clarify the TODO message or remove it

**Comment by reuvenharrison** [comment]
Hi @fenollp,
Can we do anything to help with this PR?

Please note that we expect frequent tweaks on this functionality during the coming months until it is stable, so
perhaps it would be better to implement it in https://github.com/Tufin/oasdiff to speed up the process?

Thank you,
Reuven

**Comment by fenollp** [comment]
Hi @reuvenharrison !
I believe this should be implemented as a schema Walker. So first let's add a `Walk` method so a callback func can be called by the visitor. Then impl a `MergeAllOf` func that can be given to that visitor.
This should provide a simple and extensible API for you and other consumers of this lib (and I have other use cases for that schema walker). Also it helps cut this PR in two parts.

Does this seem reasonable to you?

I don't kind this func to change in the future. I agree its impl could leave in your repo but I'd rather the lib offers this kind of thing from upstream somewhat. But if you prefer let's have the visitor in this lib and you impl your merge in your repo and once it's stable enough for you then open a second PR with your merge impl. That's probably the easiest for everyone.

**Comment by reuvenharrison** [comment]
Hi @fenollp,
Thanks for your quick reply.
We will proceed according to your suggestion above.
Regards,
Reuven

**Comment by reuvenharrison** [comment]
We implemented this in [oasdiff](https://github.com/Tufin/oasdiff) in order to answer the primary requirement ASAP:
https://github.com/Tufin/oasdiff/blob/main/ALLOF.md
 
Happy to assist integration/migration to this repo if anyone wants to lead this effort.

**Pull request closed without merging**